addons/core/addon/decorators/resource-filter.js

```diff
@@ -0,0 +1,251 @@
+import EmberObject from '@ember/object';
+import { getOwner } from '@ember/application';
+import { typeOf } from '@ember/utils';
+
+/**
+ * @param value
+ * @return {boolean} true if `value` is a function
+ */
+const isFunction = (value) => typeOf(value) === 'function';
+
+/**
+ * Represents a filter on a specific route and maintains filter state.
+ * @class
+ */
+class RouteResourceFilter extends EmberObject {
+  // =attributes
+
+  route;
+  name;
+  defaultValue;
+  allowed;
+  serialize;
+  findBySerialized;
+
+  // =computed
+
+  /**
+   * Query parameter name where the selected value(s) for this filter are
+   * stored in the URL.
+   * @type {string}
+   */
+  get filterKey() {
+    return `filter-${this.name}`;
+  }
+
+  /**
+   * The allowed values for this filter.  See `allowed`.
+   * @type {?array}
+   */
+  get allowedValues() {
```

Comment:
For all the computed functions declared here, do we plan to unit test them?

Reply:
I've been lax on this because filtering is in flux.  Testing is something I'd like to come back to in another pass.